pkg/cloudprovider/provider/kubevirt/provider.go

```diff
@@ -651,13 +709,34 @@ func (p *provider) newVirtualMachine(ctx context.Context, c *Config, pc *provide
 					TopologySpreadConstraints:     getTopologySpreadConstraints(c, map[string]string{machineDeploymentLabelKey: labels[machineDeploymentLabelKey]}),
 				},
 			},
-			DataVolumeTemplates: getDataVolumeTemplates(c, dataVolumeName),
+			DataVolumeTemplates: getDataVolumeTemplates(c, dataVolumeName, getDataVolumeSource(c, mdName)),
 		},
 	}
 	return virtualMachine, nil
 }
 
-func (p *provider) Cleanup(ctx context.Context, machine *clusterv1alpha1.Machine, _ *cloudprovidertypes.ProviderData) (bool, error) {
+func (p *provider) createPreAllocatedDataVolume(ctx context.Context, c *Config, machine *clusterv1alpha1.Machine, sigClient client.Client, dvName string) error {
+	existingDV := &cdiv1beta1.DataVolume{}
+	exist := true
```

Comment:
I think you can get rid of this flag, it's not required. You can check if the object is empty.

Reply:
What I mean here is that you can return an error from Get and implement lower a check that if err is not found then create a new data volume.